app/controllers/application_controller.rb

```diff
@@ -466,6 +468,23 @@ def better_errors_allow_inline
     raise
   end
 
+  def current_ability
+    Ability.new(current_user).merge(granted_capability)
+  end
+
+  def granted_capability
+    Capability.new(current_user, current_token)
+  end
+
+  def deny_access(_exception)
+    if current_user
```

Comment:
This should be `current_token` here I think? Otherwise we wind up talking about OAuth when it wasn't used - the current code is guarded by a token test at least.

If we did that and then had a suitably generic case after it for `current_user` then I think we could get rid of many of the special case overrides for `deny_access` in the other controllers - most of those are pretty much impossible to hit anyway so don't really need to give the precise detail they do now.

Reply:
I'm in general happy with telling people that they don't have permission (rather than what permission they need) to avoid all the special case overrides. I agree there's something not quite right on handling tokens vs current_user here.